src/coreclr/jit/morph.cpp

```diff
@@ -18133,3 +18141,80 @@ GenTreeArgList* Compiler::fgMorphArgList(GenTreeArgList* args, MorphAddrContext*
 
     return args;
 }
+
+//------------------------------------------------------------------------
+// fgMorphReduceAddOps: reduce successive variable adds into a single multiply,
+// e.g., i + i + i + i => i * 4.
+//
+// Arguments:
+//    tree - tree for reduction
+//
+// Return Value:
+//    reduced tree if pattern matches, original tree otherwise
+//
+GenTree* Compiler::fgMorphReduceAddOps(GenTree* tree)
+{
+    // ADD(_, V0) starts the pattern match.
+    if (!tree->OperIs(GT_ADD) || tree->gtOverflow())
+    {
+        return tree;
+    }
+
+#ifndef TARGET_64BIT
+    // Transforming 64-bit ADD to 64-bit MUL on 32-bit system results in replacing
+    // ADD ops with a helper function call. Don't apply optimization in that case.
+    if (tree->TypeGet() == TYP_LONG)
+    {
+        return tree;
+    }
+#endif
+
+    GenTree* op1 = tree->AsOp()->gtOp1;
+    GenTree* op2 = tree->AsOp()->gtOp2;
+
+    if (!op2->OperIs(GT_LCL_VAR) || !varTypeIsIntegral(op2))
+    {
+        return tree;
+    }
+
+    int          foldCount = 0;
+    unsigned int lclNum    = op2->AsLclVarCommon()->GetLclNum();
+
+    // Search for pattern of shape ADD(ADD(ADD(lclNum, lclNum), lclNum), lclNum).
+    while (true)
+    {
+        // ADD(lclNum, lclNum), end of tree
+        if (op1->OperIs(GT_LCL_VAR) && op1->AsLclVarCommon()->GetLclNum() == lclNum && op1->OperIs(GT_LCL_VAR) &&
+            op2->AsLclVarCommon()->GetLclNum() == lclNum)
+        {
+            foldCount += 2;
+            break;
+        }
+        // ADD(ADD(X, Y), lclNum), keep descending
+        else if (op1->OperIs(GT_ADD) && !op1->gtOverflow() && op2->OperIs(GT_LCL_VAR) &&
+                 op2->AsLclVarCommon()->GetLclNum() == lclNum)
+        {
+            foldCount++;
+            op2 = op1->AsOp()->gtOp2;
+            op1 = op1->AsOp()->gtOp1;
+        }
+        // Any other case is a pattern we won't attempt to fold for now.
+        else
+        {
+            return tree;
+        }
+    }
+
+    // V0 + V0 ... + V0 becomes V0 * foldCount, where postorder transform will optimize
+    // accordingly
+    GenTree* lclVarTree = tree->AsOp()->gtOp2;
+    GenTree* consTree   = tree->AsOp()->gtOp1;
+    consTree->BashToConst(foldCount, lclVarTree->TypeGet());
+
+    // GT_MUL operators can later be transformed into 'GT_CALL'.
+    GenTree* morphed =
+        new (this, GT_CALL) GenTreeOp(GT_MUL, tree->TypeGet(), lclVarTree, consTree DEBUGARG(/*largeNode*/ true));
```

Comment:
No longer needed (`gtNewOperNode` can be used instead)?

(This will need a comment update above, where we give up for 64 `MUL`s on 32 bit, that it is now for both CQ and correctness reasons)